torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2262,7 +2365,83 @@ def _load_sharded_state_dict(
         with self.set_state_dict_type(StateDictType.SHARDED_STATE_DICT):
             return self.load_state_dict(state_dict, strict)
 
+    def _patch_forwards(self):
+        """Saves the ``forward()`` method of each module in the wrapped module
+        hierarchy and patches them to add additional FSDP logic."""
+        self._module_to_forward.clear()
+        for module in self._fsdp_wrapped_module.modules():
+            self._module_to_forward[module] = module.forward
+            module.forward = self._patch_forward(module)
+
+    def _patch_forward(self, module: nn.Module) -> Callable:
+        """Patches ``module.forward()`` to add additional FSDP logic."""
+        def reshard(_params: List[FlatParameter]):  # TODO (awgu): refactor
+            self._free_full_params(_params)
+            if self._mixed_precision_enabled_for_params():
+                self._free_mp_shard(_params)
+            self._use_param_local_shard(_params)
+
+        module_forward = module.forward
+
+        def patched_forward(*args, **kwargs):
+            with torch.autograd.profiler.record_function("FullyShardedDataParallel.forward"):
+                self._lazy_init()
+                self.training_state = TrainingState_.FORWARD
+                # Move inputs to GPU and cast as needed
+                if self._is_root:
+                    args, kwargs = _to_kwargs(args, kwargs, self.compute_device.index, False)
+                    args = args[0]
+                    kwargs = kwargs[0]
+                    if self._mixed_precision_enabled_for_params():
+                        input_dtype = self.mixed_precision.param_dtype
+                        args, kwargs = self._cast_fp_inputs_to_precision(
+                            input_dtype, *args, **kwargs,
+                        )
+                handles = self._module_to_handles[module]
```

Comment:
I am sometimes confused with handles and self._handles. I guess for this we can just use self._module_to_handles[module] throughout within the patch_forward function, this would be clearer to me